### PR TITLE
Spawn commands like npm

### LINKED ISSUE #20: Tracking: Using spawn vs. exec as underlying process mechanism
In investigating FormidableLabs/radium#420, I'm beginning to suspect that the users' shell configuration is a contributing factor. For example, you can configure your shell in a strict mode so that if any command fails in a `cmd1 || cmd2` chain, the whole command still fails, as if they were joined with `&&`. It's possible these people have some global setting that's causing the `||` to fail (yes, even though NPM attempts to run your script with plain old `sh`).

Anyway, whether or not that is the issue, I dug into how `npm` runs scripts, and found that [they use `spawn`](https://github.com/npm/npm/blob/580f88a8183375a9c045529f10bf1aaea75f7bdd/lib/utils/lifecycle.js#L224) (they wrap Node's `spawn` slightly). In order to not have to parse the command, they run `sh -c` which takes the shell command to run as a string.

While I'm fine with `exec`, I thought I'd bring this back up because it's a difference in how Builder runs your script vs. NPM – and I anticipate that difference being a source of issues and confusion in a future where Builder is used more widely.

(We can put this off for now, just thought I'd add it as a TODO.)

### PR DESCRIPTION
**Pull request by exogen**
This fixes the failing repro case from #120, but breaks the tests. If we go this route we need to rethink how we're capturing child process output in our tests, because stubbing `process.stdout.write` doesn't work anymore – a pretty major problem for our test suite.

The magic that makes it work is the `stdio: "inherit"` option to `spawn` (this is also what breaks the tests). AFAIK using anything but `"inherit"` won't work (I tried, and could still repro the issue).

This currently just makes the existing `buffer` option a no-op, but we could easily bring it back: if `buffer` is specified we'd switch off `stdio: "inherit"` and use `stdio: "pipe"` (the default) instead, buffering the output ourselves like `exec` would do.

Tasks:

* [x] Fix tests by finding a new way to capture stdout/stderr.
* [x] Bring back `buffer` option.
* [x] Test `spawn` edge cases.
* [x] Remove `--unlimited-buffer` option and tests.

Tickets:

* Need continued work to solve #120
* Fixes #20

### CODE REVIEW

File: lib/runner.js [hunk 1]
```diff
@@ -157,15 +143,21 @@ var expandArchetype = function (cmd, opts, env) {
...
-  var maxBuffer = opts.unlimitedBuffer ? Infinity : MAX_BUFFER;
+  var sh = "sh";
+  var shFlag = "-c";
 
   // Update shell options and ensure basic structure.
   shOpts = _.extend({
-    maxBuffer: maxBuffer,
-    env: {}
+    env: {},
+    stdio: "inherit"
```
Comment (exogen) [inline lib/runner.js]: This is the magic.

File: lib/spawn.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  // The "error" event will almost never happen unless there's a problem at the
+  // OS level; things like "command not found" and non-zero exit codes will be
+  // normal "close" events, but with the appropriate `code` and `signal`.
+  proc.on("error", function (err) {
+    error = err;
+  });
+
+  proc.on("close", function (code, signal) {
+    if (error) {
+      callback(error);
```
Comment (ryan-roemer) [inline lib/spawn.js]: NIT: Switch to `return callback()` short-circuit logic.

File: lib/runner.js [hunk 3]
```diff
@@ -157,15 +143,21 @@ var expandArchetype = function (cmd, opts, env) {
...
   }, shOpts);
 
-  var buffer = opts.buffer;
+  if (process.platform === "win32") {
```
Comment (ryan-roemer) [inline lib/runner.js]: NIT: Is there a weblink for where you got this wrapper code? Or maybe just more explanation about windows shell if you're doing it from scratch?

**Comment by exogen** [comment]
Updated. At the moment, any child process stdout/stderr assertions were removed from tests. From doing some reading, apparently using `stdio: "inherit"` makes reading the child's output impossible, at least in any easy way we'd expect to use. I'm inclined to believe it's possible, at least in a roundabout or platform-specific way. For tests, we could possibly change the tasks to redirect their output to a file (or file descriptor?) and check the contents afterward.

Outside of tests, it's probable that we don't actually *need* Builder to be able to read the output of the child. After all, if npm is using this method then it can't either. If it's good enough for npm...

File: test/server/spec/base.spec.js [hunk 4]
```diff
@@ -45,4 +45,14 @@ afterEach(function () {
   base.mockFs.restore();
   base.sandbox.restore();
   log._unsetLevel();
+  try {
+    fs.unlinkSync("stdout.log");
```
Comment (ryan-roemer) [inline test/server/spec/base.spec.js]: Can you async this?

File: test/server/spec/base.spec.js [hunk 5]
```diff
@@ -45,4 +45,14 @@ afterEach(function () {
...
+  } catch (err) {
+    // No worries.
+  }
+  try {
+    fs.unlinkSync("stderr.log");
```
Comment (ryan-roemer) [inline test/server/spec/base.spec.js]: Same here

File: test/server/spec/bin/builder-core.spec.js [hunk 6]
```diff
@@ -21,24 +22,14 @@ var run = require("../../../../bin/builder-core");
...
-  return function (done) {
-    base.sandbox.stub(process.stdout, "write");
-
-    var _done = function (err) {
-      process.stdout.write.restore();
-      done(err);
-    };
+var readOut = function () {
+  base.mockFs.restore();
+  return fs.readFileSync("stdout.log", { encoding: "utf8" });
```
Comment (ryan-roemer) [inline test/server/spec/bin/builder-core.spec.js]: Can you async this?

Side Note: The sync methods when mockFS is active is fine because we're going against in-memory mocks. These appear to be real fs access once you've restored from the mock fs...

File: test/server/spec/bin/builder-core.spec.js [hunk 7]
```diff
@@ -21,24 +22,14 @@ var run = require("../../../../bin/builder-core");
...
 
-    try {
-      return fn(_done);
-    } catch (err) {
-      return _done(err);
-    }
-  };
+var readErr = function () {
+  base.mockFs.restore();
+  return fs.readFileSync("stderr.log", { encoding: "utf8" });
```
Comment (ryan-roemer) [inline test/server/spec/bin/builder-core.spec.js]: Same here.

File: test/server/spec/bin/builder-core.spec.js (hunk 6 above)
Comment (exogen) [inline test/server/spec/bin/builder-core.spec.js]: I can, it'll just make the assertions using it a lot more tedious. Is there an actual benefit? Mocha only runs one test at a time, but even if they were parallel, we never want multiple things trying to touch these files at the same time.

**Comment by exogen** [comment]
Switched to async `fs` methods wrapped in a Promise, but didn't use Mocha's return-a-Promise feature cuz the whole outer `run` would need to be wrapped; it was messy. So I kept `done` and am doing `.then(done, done)`.

File: lib/runner.js [hunk 8]
```diff
@@ -157,15 +143,23 @@ var expandArchetype = function (cmd, opts, env) {
  * @returns {Object}            Child process object
  */
 var run = function (cmd, shOpts, opts, callback) {
-  var maxBuffer = opts.unlimitedBuffer ? Infinity : MAX_BUFFER;
```
Comment (ryan-roemer) [inline lib/runner.js]: CLEANUP NOTE -- Need to remove `--unlimited-buffer` option and tests.

**Comment by exogen** [comment]
`async` refactor LGTM 👍

File: lib/runner.js [hunk 9]
```diff
@@ -173,10 +177,28 @@ var run = function (cmd, shOpts, opts, callback) {
   var proc = spawn(sh, [shFlag, cmd], shOpts, function (err) {
     var code = err ? err.code || 1 : 0;
     var level = code === 0 ? "info" : "warn";
+
+    // Output buffered output.
+    if (buffer) {
+      _.each(bufs, function (buf) {
```
Comment (exogen) [inline lib/runner.js]: Any reason not to use native `bufs.forEach`?

**Comment by exogen** [comment]
LGTM 👍
